### PR TITLE
Feature/fk bv 258

### PR DESCRIPTION
**Pull request by furkhat**

### CODE REVIEW

File: data/test.go [hunk 1]
```diff
@@ -289,13 +289,22 @@ func RollbackTestTX(t *testing.T, tx *reform.TX) {
 	}
 }
 
+// FindInTestDB selects a record from test db
+func FindInTestDB(t *testing.T, db *reform.DB,
+	str reform.Struct, column string, arg interface{}) {
+	if err := db.FindOneTo(str, column, arg); err != nil {
+		t.Fatalf("failed to find %T: %v. %s", str, err, util.Caller())
```
Comment (ababo) [inline data/test.go]: Better may be "failed to find %T: %v (%s)" (here and below).

File: uisrv/account.go [hunk 2]
```diff
@@ -184,19 +187,49 @@ func (s *Server) handleCreateAccount(w http.ResponseWriter, r *http.Request) {
...
 func (s *Server) handleUpdateAccountBalance(w http.ResponseWriter, r *http.Request, id string) {
-	// TODO: validate request params and create balance job.
-	w.WriteHeader(http.StatusBadRequest)
+	payload := &accountBalancePayload{}
+	if !s.parsePayload(w, r, payload) {
+		return
+	}
+	if payload.Amount == 0 || (payload.Destination != data.ContractPSC &&
+		payload.Destination != data.ContractPTC) {
+		w.WriteHeader(http.StatusBadRequest)
```
Comment (ababo) [inline uisrv/account.go]: No message to user?

File: uisrv/income.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+const (
+	incomeByOffering = "offering"
+	incomeByProduct  = "product"
+)
+
+func (s *Server) handleGetIncome(w http.ResponseWriter, r *http.Request) {
+	var arg, query string
+
+	if arg = r.FormValue(incomeByOffering); arg != "" {
+		query = `select sum(receipt_balance)
```
Comment (ababo) [inline uisrv/income.go]: Please use [SQL code style](https://www.sqlstyle.guide/).

File: uisrv/offering.go [hunk 4]
```diff
@@ -156,8 +156,27 @@ func (s *Server) handlePutOfferingStatus(
 	if !s.parsePayload(w, r, req) {
 		return
 	}
+	// TODO: popup, deactivate
+	if req.Action != PublishOffering {
+		w.WriteHeader(http.StatusBadRequest)
+		return
```
Comment (ababo) [inline uisrv/offering.go]: No message?

File: uisrv/offering.go [hunk 5]
```diff
@@ -156,8 +156,27 @@ func (s *Server) handlePutOfferingStatus(
...
+	if err := s.queue.Add(&data.Job{
+		Type:        data.JobAgentPreOfferingMsgBCPublish,
+		RelatedType: data.JobOfferring,
+		RelatedID:   id,
+		CreatedBy:   data.JobUser,
+		Data:        []byte("{}"),
+	}); err != nil {
+		s.logger.Error("failed to add %s: %v",
+			data.JobAgentPreOfferingMsgBCPublish, err)
+		s.replyUnexpectedErr(w)
```
Comment (ababo) [inline uisrv/offering.go]: It makes sense to add `return` even if it's a last function statement.

File: uisrv/offering_test.go [hunk 6]
```diff
@@ -320,8 +320,33 @@ func getOfferingStatus(t *testing.T, id string) *http.Response {
 	return sendToOfferingStatus(t, id, "", "GET")
 }
 
+func sendOfferingAction(t *testing.T, id, action string) *http.Response {
+	path := fmt.Sprint(offeringsPath, id, "/status")
```
Comment (ababo) [inline uisrv/offering_test.go]: Maybe to make a path constant?

Comment (furkhat) [reply uisrv/offering_test.go]: `id` is generic

**Pull request merged**